### PR TITLE
Fix lookup matching from ISIS Reflectometry Preview tab

### LINKED ISSUE #35704: Crash when load run on Preview Tab that matches multiple angle settings
**Original reporter:** Max at ISIS

**Describe the bug**

Loading a run on the preview tab with an angle that matches multiple rows from the lookup table on the Experiment Settings tab causes Mantid to crash. For the normal batch reduction we match on angle and title, however on the preview tab the matching is done on angle alone because there is no title information provided via the GUI. When this matches against multiple entries from the lookup table it causes a `MultipleRowsFoundException` that isn't being handled, which causes Mantid to crash.

**To Reproduce**

We have a batch file we can use for testing with run number 69252, which is how the error was originally discovered. The bug can also be reproduced more simply as follows:

1) Open the ISIS Reflectometry GUI.
1) Go to the Experiment Settings tab and add two new rows to the lookup settings table - one with angle 2.3 and no title and one with angle 2.3 and a title.
1) Go to the Reduction Preview tab and enter run `INTER45455` and angle `2.3`. Click Load. Mantid will crash.

**Expected behavior**

Mantid should not crash if there are issues with finding lookup-table information, so we should improve the error handling. An example of handling this same error on the Runs tab can be seen in `BatchJobManager::addAlgorithmForProcessingRow`.

I have confirmed with our scientists that they would like the preview tab to be matching on angle plus title in the same way as is done via the normal batch reduction. This means taking the group name/title part of the `run_title` information (everything up to the `th=` section) and using that for the matching. Note method `SearchResult::parseMetadataFromTitle` appears to contain the logic that extracts the group name on the Runs tab.

On the Runs tab it is possible to edit the group name that has been extracted from the run title, however Max has confirmed that this is almost never changed so they do not need to be able to do this via the preview tab. It would, however, be helpful to see the run title (the full title, not the part only taken for matching) on the preview tab. Either at the top next to the load controls or above the plot.

If we change the matching logic then we should also check that the `Apply` button on the preview tab is still setting the selected regions on the correct lookup row.

**Platform/Version (please complete the following information):**
 - Mantid Version: since version 6.6

### LINKED ISSUE #35825: Preview tab crashes if reduction is triggered with no loaded workspace
**Describe the bug**

During testing for another issue I noticed that it is possible to trigger a reduction on the preview tab when no workspace is loaded. This can happen when the update angle button is clicked, or if a run is successfully loaded and then there is an error trying to load a different run.

**To Reproduce**

To reproduce with the update angle button:

1) Open the ISIS Reflectometry interface
1) Go to the Reduction Preview tab
1) Enter `POLREF37125` into the `Run` field and click the `Load` button (multi-period data is not supported by the preview tab at the moment so it will cause a loading error)
1) Change the angle in the spin box
1) Click the `Update` button. Mantid should crash, or load a dialog with a failed assert and this eventually leads to a crash.

To reproduce with changing slice viewer regions:

1) Open the ISIS Reflectometry interface
1) Go to the Experiment Settings tab and select `MultiDetectorAnalysis` from the `AnalysisMode` dropdown.
1) Go to the Reduction Preview tab
1) Enter `INTER61135` into the `Run` field and click the `Load` button
1) Select a region on the slice viewer plot
1) Enter `POLREF37125` into the `Run` field and click the `Load` button
1) Move the existing region on the slice viewer plot. Mantid should crash, or load a dialog with a failed assert and this eventually leads to a crash.

**Expected behavior**

Mantid should not crash. Given that this is quite a niche issue I think simply preventing the crash and displaying an appropriate error message is sufficient.

**Platform/Version (please complete the following information):**
 - Mantid Version - definitely 6.7, but believe since 6.6

### PR DESCRIPTION
**Pull request by rbauststfc**
**Description of work**

**Purpose of work**

This is part two of the fix for the linked issue. This PR adds the functionality to match against lookup rows using both run title and angle, rather than just angle. During testing I also found a crash that occurs when a reduction is triggered without the loaded workspace, so this PR also implements a simple fix for that.

**Summary of work**

This PR moves the logic for extracting the title and theta values from the logs `run_title` entry out of `SearchResult` and into `ParseReflectometryStrings`. This was to allow the logic to be re-used in the `LookupTable` class for finding the title to match against from a `PreviewRow`.

I have also changed the lookup row matching logic in `LookupTable` so that both a `Row` and `PreviewRow` share the same logic and just retrieve the title for matching in a different way.

The full run title is now displayed at the top of the Reduction Preview tab, as requested by our scientists. I have used a label to display this information, as our scientists do not need to edit it and it seemed to give the best behaviour in terms of resizing and word wrap.

Finally I have also included a small change to check if a loaded workspace is available in `PreviewPresenter::runReduction` to fix the crash described in #35825.

**To test:**

- See instructions in the relevant issues.

- For #35704, the angle and title matching can be tested by selecting a region on the instrument view and slice viewer (this will need to be done to get the reduction to complete successfully anyway) and then clicking the `Apply` button at the bottom right of the dialog. This should set the `ROI` and `ROI Detector IDs` values against the correct row in the lookup table on the Experiment Settings tab. To thoroughly test the matching you will need to try different title values - i.e. using the title "seg" in the Experiment Settings lookup row will match the title of run `INTER45455`, while something like "test" will not.
 
Fixes #35704 and fixes #35825. 

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by sf1919 — COMMENTED** [review]
I can confirm that Mantid no longer crashes. However when testing for #35704 using the instructions in the original issue and those above I found I am still getting error messages. Reading through the issues I get the impression that is what should be happening (at least at the moment) - it's just stopping it crashing. Is that correct? Error messages below for info
Screenshot (314)

File: qt/scientific_interfaces/ISISReflectometry/Reduction/LookupTable.cpp [hunk 1]
```diff
@@ -21,37 +23,48 @@ constexpr double EPSILON = std::numeric_limits<double>::epsilon();
...
   // Now filter by angle; it should be unique
-  if (auto found = searchByTheta(lookupRows, row.theta(), tolerance)) {
-    return found;
-  }
-  // If we didn't find a lookup row where theta matches, then we allow the user to specify a "wildcard" row
-  // which will be used for everything where a specific match is not found
-  auto result = findWildcardLookupRow();
-  return result;
-}
-
```
[the remark this answers is not in the document]
Comment (rbauststfc) [reply qt/scientific_interfaces/ISISReflectometry/Reduction/LookupTable.cpp:52]: I think the diff is making this look a bit confusing - we still check for the wildcard row, but that's now happening on lines 72-73.

To explain in more detail, we previously had two methods called `LookupTable::findLookupRow`, one for finding a matching lookup-row for a `Row` object and one for doing this with a `PreviewRow` object. Both these methods included the `searchByTheta` and  `findWildcardLookupRow` steps, but only the version for the `Row` object included filtering by title.

Now we're trying to perform the same steps for both (i.e. filter by title, match by theta and check wildcard row), so the code still has the two separate `LookupTable::findLookupRow` methods, one for each object type, but these both just include a first step to get the title and then make a call to a third new method, `LookupTable::findLookupRow(std::string const &title, boost::optional<double> const &theta, double tolerance)`. This new method runs through the rest of the logic that is now exactly the same for both, and so avoids duplication.

**Comment by rbauststfc** [comment]
> I can confirm that Mantid no longer crashes. However when testing for #35704 using the instructions in the original issue and those above I found I am still getting error messages. Reading through the issues I get the impression that is what should be happening (at least at the moment) - it's just stopping it crashing. Is that correct?

That error message is what I would expect to see if you just follow the instructions on #35704, so that's all working as expected for the first part of the testing.

There is then this second paragraph in the test instructions (in this PR description) for getting the reduction to complete successfully and test that the lookup table matching is working correctly:

> For #35704, the angle and title matching can be tested by selecting a region on the instrument view and slice viewer (this will need to be done to get the reduction to complete successfully anyway) and then clicking the Apply button at the bottom right of the dialog. This should set the ROI and ROI Detector IDs values against the correct row in the lookup table on the Experiment Settings tab. To thoroughly test the matching you will need to try different title values - i.e. using the title "seg" in the Experiment Settings lookup row will match the title of run INTER45455, while something like "test" will not.

So, after following the instructions on the issue, if you then select a region from the instrument view plot on the preview tab (you can just set a region around the entire plot) then the reduction should complete successfully with no further errors. You can then test selecting and applying regions from the instrument view and slice viewer plot to check that the correct lookup table row is updated (which will show that the matching is working correctly).

**Review by sf1919 — APPROVED** [review]
I have tested the functionality of this PR using all the test instructions provided. My queries regarding the outputs of that and my review of the code have been answered. I am happy to approve.

**Pull request merged**